[android] handle place page direction arrow more reliably

arnaudvergnet opened the pull request:
Following https://github.com/organicmaps/organicmaps/pull/3544, I found the place page peek height was wrong in case the title was short enough to fit in one line without the direction arrow (position not known), but too long to fit when the direction is shown. As this arrow would blink when the sheet was opened, the logic would calculate a peek height without taking into account this arrow, and thus was not enough.

This PR aims to fix this issue by more reliably displaying the direction arrow. I also slightly modified the UI to place the arrow on top of the distance and centered, felt it was prettier. 

Going from no position to position known while the pp is open. The peek height is now correctly updated.

https://user-images.githubusercontent.com/80701113/202572365-fbecd838-3f24-4410-8817-fa74abb08c13.mp4

Changing POIs when the location is known. The direction arrow is always shown, no more ugly blinking in.

https://user-images.githubusercontent.com/80701113/202572374-ddd36de8-19a3-4cb6-9e7c-797011541eb1.mp4

Mr-Mime commented:
This also fixes https://github.com/organicmaps/organicmaps/issues/3805 IMO.

I started to look into this issue and saw that the layout positions the arrow and distance on the bottom.
Thus there is a lot of white space above, which stays empty and is not used by the place name.
![2022-11-16-224532_escrotum](https://user-images.githubusercontent.com/37423773/202657611-04ab8b96-47d3-41e9-ab14-f2b297d861f3.png)

Based on the fact that the arrow is now above the distance and centered on the vertical axis, there is now more room for the name.
As it can be seen in the videos, the line wrap now looks more "natural" or "logical".

biodranik commented:
Thanks! Wouldn't it be better to take the whole upper-right space for the long (and important for users!) title, instead of the less important arrow with distance, which can be placed under the title on the right (centered)?

Mr-Mime commented:
Something like this might be good.

![mockup_layout](https://user-images.githubusercontent.com/37423773/202674680-35085c9c-68be-4335-bb27-10eafcb93ebc.jpg)

arnaudvergnet commented:
It would work only if the place has a subtitle and the download size, which is never the case when you have downloaded the map. If you look at the examples in the video, none of the places I checked would have space to put the title above the arrow.

Doing something like this would require thinking a bit more about the design, but this PR was intended to be a hotfix. Maybe we could look more into this when we decide to refactor the place page.

biodranik commented on android/src/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -320,6 +321,14 @@ protected void onFinishInflate()
+      final int oldHeight = oldBottom - oldTop;
+      final int newHeight = bottom - top;
+      if (oldHeight != newHeight)
+      {
+        mOnPlacePageContentChangeListener.OnPlacePageContentChange();
+      }
```
```suggestion
        mOnPlacePageContentChangeListener.OnPlacePageContentChange();
```
We don't use bracers for one-liners.

biodranik commented on android/src/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -194,6 +193,8 @@ public class PlacePageView extends NestedScrollViewClickFixed
   private CountryItem mCurrentCountry;
   private boolean mScrollable = true;
 
+  private double mLastKnownNorthAzimuth = 0;
```
Why do you think it is needed? What is the benefit of showing some potentially (very) outdated azimuth? Wouldn't it confuse users, for example, when POI was selected, and location updates were _disabled_ in the system settings?

arnaudvergnet replied:
The azimuth is only displayed if the location is available, so this case should no happen.

biodranik replied:
What if user disables location in the system while using OM with an open place page?

arnaudvergnet replied:
The arrow and distance is still shown, but if the user closes and reopens the place page, they are properly hidden. On current master the behavior is the same but only the distance is shown, which is still wrong.

IMO even if this is not the best way to handle this (but it would need a bigger refactoring), this is still better than what we currently have in master as only showing the distance without the arrow feels like a bug.

biodranik replied:
There likely was good reasoning behind it: without a gyroscope, you don't know the direction. But in most cases, you can't move too fast to make the distance outdated. So it makes sense even if the location was _just recently_ disabled. Of course, if the location is _too_ outdated, then the distance should not be displayed.

arnaudvergnet replied:
I can try to disable both distance and direction when the location is disabled/lost, instead of keeping the old behavior, but it might be difficult.

biodranik replied:
My point was to remove direction, but leave the distance, if there is no fresh GPS data.

biodranik reviewed (commented):
Does it look/work ok also in the dark mode and on tablets?

biodranik commented:
> but this PR was intended to be a hotfix. 

Ok, we can always improve it later.

arnaudvergnet commented:
Updated the PR to get the old behavior back.

I also fixed an issue where the place page peek height would not update when the preview content would change size (for example when the location is detected after the place page was open, making the title take two lines instead of one). In my previous refactoring I removed such a call to improve performance. But in this case, the listener is only on the preview and not on the whole bottomsheet, so the changes are less frequent.

biodranik reviewed (approved):
@rtsisyk LGTM
